src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -451,49 +459,119 @@ export class SearchSource {
   }
 
   private flatten() {
+    const { getConfig } = this.dependencies;
     const searchRequest = this.mergeProps();
 
     searchRequest.body = searchRequest.body || {};
-    const { body, index, fields, query, filters, highlightAll } = searchRequest;
+    const { body, index, query, filters, highlightAll } = searchRequest;
     searchRequest.indexType = this.getIndexType(index);
 
-    const computedFields = index ? index.getComputedFields() : {};
-
-    body.stored_fields = computedFields.storedFields;
-    body.script_fields = body.script_fields || {};
-    extend(body.script_fields, computedFields.scriptFields);
-
-    const defaultDocValueFields = computedFields.docvalueFields
-      ? computedFields.docvalueFields
-      : [];
-    body.docvalue_fields = body.docvalue_fields || defaultDocValueFields;
-
-    if (!body.hasOwnProperty('_source') && index) {
-      body._source = index.getSourceFiltering();
+    // get some special field types from the index pattern
+    const { docvalueFields, scriptFields, storedFields } = index
+      ? index.getComputedFields()
+      : {
+          docvalueFields: [],
+          scriptFields: {},
+          storedFields: ['*'],
+        };
+
+    const fieldListProvided = !!body.fields;
+    const getFieldName = (fld: string | Record<string, any>): string =>
+      typeof fld === 'string' ? fld : fld.field;
+
+    // set defaults
+    let fieldsFromSource = searchRequest.fieldsFromSource || [];
+    body.fields = body.fields || [];
+    body.script_fields = Object.assign(body.script_fields || {}, scriptFields);
+    body.stored_fields = storedFields;
+
+    // apply source filters from index pattern if specified by the user
+    let filteredDocvalueFields = docvalueFields;
+    if (index) {
+      const sourceFilters = index.getSourceFiltering();
+      if (!body.hasOwnProperty('_source')) {
+        body._source = sourceFilters;
+      }
+      if (body._source.excludes) {
+        const filter = fieldWildcardFilter(
+          body._source.excludes,
+          getConfig(UI_SETTINGS.META_FIELDS)
+        );
+        // also apply filters to provided fields & default docvalueFields
+        body.fields = body.fields.filter((fld: SearchFieldValue) => filter(getFieldName(fld)));
+        fieldsFromSource = fieldsFromSource.filter((fld: SearchFieldValue) =>
+          filter(getFieldName(fld))
+        );
+        filteredDocvalueFields = docvalueFields.filter((fld: SearchFieldValue) =>
+          filter(getFieldName(fld))
+        );
+      }
     }
 
-    const { getConfig } = this.dependencies;
+    // specific fields were provided, so we need to exclude any others
+    if (fieldListProvided || fieldsFromSource.length) {
+      const bodyFieldNames = body.fields.length
+        ? body.fields.map((field: string | Record<string, any>) => getFieldName(field))
+        : [];
```

Comment:
:art: Codestyle nitpick: Should not require a check for length, since `map` on an empty array will just result in the same empty array we're setting in the `else` here.

```suggestion
      const bodyFieldNames = body.fields.map((field: string | Record<string, any>) => getFieldName(field));
```